dmoj/judge.py

```diff
@@ -80,51 +83,37 @@ def update_problems(self):
         """
         self.updater_signal.set()
 
-    def _block_and_grade(self, problem, language, source, short_circuit, report=print):
-        if 'signature_grader' in problem.config:
-            grader_class = graders.SignatureGrader
-        elif 'interactive' in problem.config:
-            grader_class = graders.BridgedInteractiveGrader
-        elif 'custom_judge' in problem.config:
-            grader_class = graders.CustomGrader
-        else:
-            grader_class = graders.StandardGrader
-
+    def _run_grading_thread(
+        self, id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta, report
+    ):
         try:
-            self.current_grader = grader_class(self, problem, language, utf8bytes(source))
-        except CompileError as compilation_error:
-            error = compilation_error.args[0] or b'compiler exited abnormally'
-
-            report(ansi_style('#ansi[Failed compiling submission!](red|bold)'))
-            report(error.rstrip())  # don't print extra newline
-
-            self.packet_manager.compile_error_packet(error)
-            return
-        else:
-            binary = self.current_grader.binary
-            if hasattr(binary, 'warning') and binary.warning is not None:
-                self.packet_manager.compile_message_packet(binary.warning)
-
-        self.packet_manager.begin_grading_packet(self.current_grader.is_pretested)
-
-        batch_counter = 1
-        in_batch = False
+            report(
+                ansi_style(
+                    'Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...' % (problem_id, id, language)
+                )
+            )
 
-        # cases are indexed at 1
-        case_number = 1
-        try:
-            for result in self.grade_cases(
-                self.current_grader, self.current_grader.cases(), short_circuit=short_circuit
-            ):
-                if isinstance(result, BatchBegin):
-                    self.packet_manager.batch_begin_packet()
-                    report(ansi_style("#ansi[Batch #%d](yellow|bold)" % batch_counter))
-                    in_batch = True
-                elif isinstance(result, BatchEnd):
-                    self.packet_manager.batch_end_packet()
-                    batch_counter += 1
-                    in_batch = False
-                else:
+            # TODO(tbrindus): in what ways can these functions fail? Do we properly throw some IE if they do?
+            self.current_judge_process = SubmissionJudgeProcess(
+                problem_id, language, source, time_limit, memory_limit, short_circuit, meta
+            )
+            in_batch = False
+            for (type, data) in self.current_judge_process.run():
```

Comment:
Also... what if the sub process gives us garbage data that's invalid? Like say, ending the grading twice? Do we just let the bridge go wtf and disconnect us?

Reply:
Applied suggestion, since I don't mind either way.

I don't think we should do validation ourselves, and let the bridge deal with it as we always have. In a way our pipe is almost a direct pipe to the network, after all.